### DISCUSSION TITLE
Question about outputs of trace_route

### DISCUSSION THREAD

**Discussion by IvanLiuTW** [valhalla/valhalla#4026] [Q&A]
Hi, 

Before I describe my question, I'd like to say thank you to the Valhalla developers and contributors. Valhalla is amazing. 🙂

I used the map matching service, and when I examined the output, I found it was different from what the document described.

The snippet of my code
```Python
# df is the GPS data
df2 = df.loc[:, ["longitude", "laititude", "timeStamp"]]
df2.columns = ["lon", "lat", "time"]
df2['type'] = pd.Series(['via']*len(df2))

# Meili is a API service for map-matching
meili_coordinates = df2.to_json(orient='records')

# Providing needed data for the body of Meili's request
# Shape is the coordinates
meili_head = '{"shape":'
# The parameters that can be changed according to the Meili's documentation
meili_tail = ""","search_radius": 150, "shape_match":"map_snap", "costing":"bicycle", "format":"osrm"}"""

# Combining all the string into a single request
meili_request_body = meili_head + meili_coordinates + meili_tail

# The URL of the local Valhalla server
# "trace_route" means that we're sening a request to Meili
url = "http://localhost:8002/trace_route" 

# Providing headers to the request
headers = {'Content-type' : 'application/json'}

# We need to send our JSON as a string
data = str(meili_request_body)

# Sending a request
r = requests.post(url, data=data, headers=headers)
response_text = json.loads(r.text)
print(f"Status Code: {r.status_code}")

response_text.keys()
# dict_keys(['matchings', 'tracepoints', 'code'])

# Dig deep into matchings
resp = json.dumps(response_text['matchings'])
# Reading our JSON to a Pandas DataFrame
df_match = pd.read_json(resp)
df_match.head()
legs = pd.read_json(json.dumps(df_match.legs[0]))
steps = pd.read_json(json.dumps(legs.steps[0]))
steps['coords'] = steps.geometry.apply(lambda x : decode(x))

```
The `code` is ok, and I think `tracepoints` are the data of the matched points.

I was wondering what information inside `matchings`, so I took a closer look what was inside.

The pictures are the data frames of `df_match`, `legs`, `steps`
`df_match`
![image](https://user-images.githubusercontent.com/88217916/225527127-6cdc30c1-6d78-4341-a8d4-08bf1f44dc6f.png)

`legs`
![image](https://user-images.githubusercontent.com/88217916/225527180-42bbc537-adde-40ab-ab1e-77917c67036a.png)

`steps`
![image](https://user-images.githubusercontent.com/88217916/225527257-3797c42d-6736-441a-9ae8-a26e7f7e0026.png)

Since I also tried `trace_attributes` service, I assumed the `geometry` is the same as what I got from one of the outputs `shape`.

However, I don't know what `driving_side` represented, because I rode on the right side of the street. And how `weight` is calculated, and what it means.

I looked them up in the API Doc as well as the discussions, but I didn't find much explanation about them. 

Thanks for reading and answering my question. I appreciate your help.

**Comment by nilsnolde** [comment]
I personally don’t really know what you’re asking here. What is „weight“? That’s not a Valhalla attribute, so I guess you mean „cost“? That’s how the routing algorithm ended up costing your path, taking into consideration all the costing options you provide or which are default for the particular costing. Driving side is very likely the driving side in the country you did this trip in. Taiwan is left driving?


Generally: isn’t all the code only noise in this context? Best to paste a curl, that’s reproducible and easily understood. Also please don’t take screenshots of code, that’s very unpleasant to look at.

**Reply by IvanLiuTW** [reply]
Hi Nils, 

Thanks for reply. 

I am a little surprised you said weight is not a Valhalla attribute, but it is the output I got from the API request. I didn't change anything.

I apologize for pasting pictures of my dataframe for I am still quite inexperienced in writing Github content. I looked up what curl is, but I am afraid I would still use the wrong way to share the code again. Could you please suggest a tutorial for creating reproducible code?

Thank you.

**Reply by nilsnolde** [reply]
Ah you're using osrm format.. It's not in the valhalla format, but regardless it's cost, so my comment is still valid.

**Reply by IvanLiuTW** [reply]
Hi Nils,

Got it, thanks for the explanation.

Regarding the driving side, I understand your point, but Taiwan is right driving. Shall I look for OSRM format for more details?

**Reply by kevinkreiser** [reply]
when you build tiles yourself and forget to use the admin database I think you end up with left hand driving. if thats the case and im remembering right then we have the admin database around in the github issues somewhere. we should link it in the Readme for people to use. it's old but country borders don't change that frequently so it's still useful.

**Reply by IvanLiuTW** [reply]
Hi Kevin,

Thanks for reply.

I understand and I remember I built the tile by adding environment variable `tile_urls` in the command to run docker image. I am not sure if this way is building myself or using the admin database.

Thanks again for your explanation.
